### PR TITLE
ARROW-12656: [C++][Gandiva] Implement castVARCHAR for date, intervalDay and intervalYear

### PR DESCRIPTION
**Pull request by rodrigojdebem**

### CODE REVIEW

**Comment by pprudhvi** [comment]
castVARCHAR_intervalyear is missing, could you please add the impl or edit the commit message above?

**Comment by anthonylouisbsb** [comment]
I add in this PR the implementation for the `castVARCHAR(time, len)` function in Gandiva

**Review by anthonylouisbsb — CHANGES_REQUESTED** [review]
The castVarchar function for date is already implemented in Gandiva. Checks the difference between the implementation.

File: cpp/src/gandiva/function_registry_datetime.cc [hunk 1]
```diff
@@ -65,8 +65,16 @@ std::vector<NativeFunction> GetDateTimeFunctionRegistry() {
                      kResultNullIfNull, "castTIMESTAMP_utf8",
                      NativeFunction::kNeedsContext | NativeFunction::kCanReturnErrors),
 
-      NativeFunction("castVARCHAR", {}, DataTypeVector{timestamp(), int64()}, utf8(),
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/function_registry_datetime.cc:68]: The error is due to this removed function

**Review by anthonylouisbsb — CHANGES_REQUESTED** [review]

**Review by anthonylouisbsb — APPROVED** [review]

**Review by ayushpranjal-dremio — COMMENTED** [review]
Can you also add an integration test (eg: a projector test)?

File: cpp/src/gandiva/precompiled/time.cc [hunk 2]
```diff
@@ -920,6 +920,202 @@ const char* castVARCHAR_timestamp_int64(gdv_int64 context, gdv_timestamp in,
   return ret;
 }
```
Comment (ayushpranjal-dremio) [inline cpp/src/gandiva/precompiled/time.cc:922]: Can you add a short description of what this function is supposed to do?

Comment (ViniciusSouzaRoque) [reply cpp/src/gandiva/precompiled/time.cc:960]: Done

Comment (ViniciusSouzaRoque) [reply cpp/src/gandiva/precompiled/time.cc:960]: Done

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Pull request closed without merging**